File: step_10_sliding_window/10.1_medium/P004_fruits_into_baskets/Solution.java
/**
 * Problem: Fruits into Baskets (LeetCode #904)
 * Difficulty: MEDIUM | XP: 25
 *
 * Find the longest subarray with at most 2 distinct values.
 * Key insight: Sliding window + frequency map with K=2 distinct limit.
 *
 * @author DSA_Nova
 */
import java.util.*;

// ============================================================
// Approach 1: Brute Force (Check All Subarrays)
// Time: O(n^2) | Space: O(1)
// ============================================================
class BruteForce {
    public static int totalFruit(int[] fruits) {
        int n = fruits.length;
        int ans = 0;
        for (int i = 0; i < n; i++) {
            Set<Integer> types = new HashSet<>();
            for (int j = i; j < n; j++) {
                types.add(fruits[j]);
                if (types.size() > 2) break;
                ans = Math.max(ans, j - i + 1);
            }
        }
        return ans;
    }
}

// ============================================================
// Approach 2: Optimal (Sliding Window -- Shrinking)
// Time: O(n) | Space: O(1)
// ============================================================
class Optimal {
    public static int totalFruit(int[] fruits) {
        int n = fruits.length;
        Map<Integer, Integer> freq = new HashMap<>();
        int left = 0, ans = 0;

        for (int right = 0; right < n; right++) {
            freq.merge(fruits[right], 1, Integer::sum);

            // Shrink until at most 2 distinct types
            while (freq.size() > 2) {
                int leftFruit = fruits[left];
                freq.merge(leftFruit, -1, Integer::sum);
                if (freq.get(leftFruit) == 0) {
                    freq.remove(leftFruit);
                }
                left++;
            }
            ans = Math.max(ans, right - left + 1);
        }
        return ans;
    }
}

// ============================================================
// Approach 3: Best (Sliding Window -- Non-Shrinking)
// Time: O(n) | Space: O(1)
// ============================================================
class Best {
    public static int totalFruit(int[] fruits) {
        int n = fruits.length;
        Map<Integer, Integer> freq = new HashMap<>();
        int left = 0;

        for (int right = 0; right < n; right++) {
            freq.merge(fruits[right], 1, Integer::sum);

            // If invalid, slide window by 1 (don't shrink)
            if (freq.size() > 2) {
                int leftFruit = fruits[left];
                freq.merge(leftFruit, -1, Integer::sum);
                if (freq.get(leftFruit) == 0) {
                    freq.remove(leftFruit);
                }
                left++;
            }
        }
        return n - left;
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Fruits into Baskets ===\n");

        int[][] inputs = {
            {1, 2, 1},
            {0, 1, 2, 2},
            {1, 2, 3, 2, 2},
            {3, 3, 3, 1, 2, 1, 1, 2, 3, 3, 4},
            {1},
            {1, 1, 1, 1},
            {1, 2, 3, 4, 5}
        };
        int[] expected = {3, 3, 4, 5, 1, 4, 2};

        for (int t = 0; t < inputs.length; t++) {
            int[] fruits = inputs[t];
            int b = BruteForce.totalFruit(fruits);
            int o = Optimal.totalFruit(fruits);
            int r = Best.totalFruit(fruits);
            boolean pass = (b == expected[t] && o == expected[t] && r == expected[t]);

            System.out.println("Input:    " + Arrays.toString(fruits));
            System.out.println("Brute:    " + b);
            System.out.println("Optimal:  " + o);
            System.out.println("Best:     " + r);
            System.out.println("Expected: " + expected[t]);
            System.out.println("Pass:     " + pass);
            System.out.println();
        }
    }
}

File: step_10_sliding_window/10.1_medium/P004_fruits_into_baskets/solution.py
"""
Problem: Fruits into Baskets (LeetCode #904)
Difficulty: MEDIUM | XP: 25

Find the longest subarray with at most 2 distinct values.
Key insight: Sliding window + frequency map with K=2 distinct limit.
"""
from typing import List
from collections import defaultdict


# ============================================================
# APPROACH 1: BRUTE FORCE (Check All Subarrays)
# Time: O(n^2) | Space: O(1)
# ============================================================
def brute_force(fruits: List[int]) -> int:
    """Try every starting index, extend as far as possible with <= 2 types."""
    n = len(fruits)
    ans = 0
    for i in range(n):
        types = set()
        for j in range(i, n):
            types.add(fruits[j])
            if len(types) > 2:
                break
            ans = max(ans, j - i + 1)
    return ans


# ============================================================
# APPROACH 2: OPTIMAL (Sliding Window -- Shrinking)
# Time: O(n) | Space: O(1)
# ============================================================
def optimal(fruits: List[int]) -> int:
    """Sliding window: expand right, shrink left when > 2 distinct types."""
    freq = defaultdict(int)
    left = 0
    ans = 0

    for right in range(len(fruits)):
        freq[fruits[right]] += 1

        # Shrink until at most 2 distinct types
        while len(freq) > 2:
            freq[fruits[left]] -= 1
            if freq[fruits[left]] == 0:
                del freq[fruits[left]]
            left += 1

        ans = max(ans, right - left + 1)

    return ans


# ============================================================
# APPROACH 3: BEST (Sliding Window -- Non-Shrinking)
# Time: O(n) | Space: O(1)
# ============================================================
def best(fruits: List[int]) -> int:
    """Non-shrinking window: slide by 1 when invalid, never shrink below previous best."""
    n = len(fruits)
    freq = defaultdict(int)
    left = 0

    for right in range(n):
        freq[fruits[right]] += 1

        # If invalid, slide window by 1
        if len(freq) > 2:
            freq[fruits[left]] -= 1
            if freq[fruits[left]] == 0:
                del freq[fruits[left]]
            left += 1

    return n - left


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Fruits into Baskets ===\n")

    test_cases = [
        ([1, 2, 1], 3),
        ([0, 1, 2, 2], 3),
        ([1, 2, 3, 2, 2], 4),
        ([3, 3, 3, 1, 2, 1, 1, 2, 3, 3, 4], 5),
        ([1], 1),
        ([1, 1, 1, 1], 4),
        ([1, 2, 3, 4, 5], 2),
    ]

    for fruits, expected in test_cases:
        b = brute_force(fruits)
        o = optimal(fruits)
        r = best(fruits)
        status = "PASS" if b == expected and o == expected and r == expected else "FAIL"
        print(f"Input: {fruits}")
        print(f"  Brute:    {b}")
        print(f"  Optimal:  {o}")
        print(f"  Best:     {r}")
        print(f"  Expected: {expected}  [{status}]\n")
